src/Microsoft.Net.Http.Headers/MediaTypeHeaderValue.cs

```diff
@@ -170,55 +212,148 @@ public IList<NameValueHeaderValue> Parameters
         public StringSegment SubType
         {
             get
             {
-                return _mediaType.Subsegment(_mediaType.IndexOf('/') + 1);
+                return _mediaType.Subsegment(_mediaType.IndexOf(ForwardSlashCharacter) + 1);
             }
         }
 
         /// <summary>
-        /// MediaType = "*/*"
+        /// Gets subtype of the <see cref="MediaTypeHeaderValue"/>, excluding any structured syntax suffix. Returns <see cref="StringSegment.Empty"/>
+        /// if there is no subtype without suffix.
         /// </summary>
-        public bool MatchesAllTypes
+        /// <example>
+        /// For the media type <c>"application/vnd.example+json"</c>, the property gives the value
+        /// <c>"vnd.example"</c>.
+        /// </example>
+        public StringSegment SubTypeWithoutSuffix
         {
             get
             {
-                return MediaType.Equals("*/*", StringComparison.Ordinal);
+                var startOfSuffix = _mediaType.LastIndexOf(PlusCharacter);
+                var startOfSubType = _mediaType.IndexOf(ForwardSlashCharacter) + 1;
+                if (startOfSuffix == -1)
+                {
+                    return _mediaType.Subsegment(startOfSubType);
+                }
+                return _mediaType.Subsegment(startOfSubType, startOfSuffix - startOfSubType);
             }
         }
 
         /// <summary>
-        /// SubType = "*"
+        /// Gets the structured syntax suffix of the <see cref="MediaTypeHeaderValue"/> if it has one.
+        /// See <see href="https://tools.ietf.org/html/rfc6838#section-4.8">The RFC documentation on structured syntaxes.</see>
         /// </summary>
-        public bool MatchesAllSubTypes
+        /// <example>
+        /// For the media type <c>"application/vnd.example+json"</c>, the property gives the value
+        /// <c>"json"</c>.
+        /// </example>
+        public StringSegment SubTypeSuffix
         {
             get
             {
-                return SubType.Equals("*", StringComparison.Ordinal);
+                var startOfSuffix = _mediaType.LastIndexOf(PlusCharacter);
+                if (startOfSuffix == -1)
+                {
+                    return default(StringSegment);
+                }
+                return _mediaType.Subsegment(startOfSuffix + 1);
             }
         }
 
+
+        /// <summary>
+        /// Get a <see cref="IList{T}"/> of facets of the <see cref="MediaTypeHeaderValue"/>. Facets are a
+        /// period separated list of StringSegments in the <see cref="SubTypeWithoutSuffix"/>.
+        /// See <see href="https://tools.ietf.org/html/rfc6838#section-3">The RFC documentation on facets.</see>
+        /// </summary>
+        /// <example>
+        /// For the media type <c>"application/vnd.example+json"</c>, the property gives the value:
+        /// <c>{"vnd", "example"}</c>
+        /// </example>
+        public IList<StringSegment> Facets
```

Comment:
Should we do custom enumerators for this? I'm just bringing it up cuz is one of the things we did for perfz on MVC. @rynowak